service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/auth/IotHubServiceSasToken.java

```diff
@@ -37,37 +37,61 @@
-     * the specified amount of time. (1 year specified in TOKEN_VALID_SECS)
+     * the specified amount of time. (1 year specified in DEFAULT_TOKEN_LIFESPAN_SECONDS)
      *
      * @param iotHubConnectionString Connection string object containing the connection parameters
      */
     public IotHubServiceSasToken(IotHubConnectionString iotHubConnectionString)
     {
-        // Codes_SRS_SERVICE_SDK_JAVA_IOTHUBSERVICESASTOKEN_12_001: [The constructor shall throw IllegalArgumentException if the input object is null]
+        this(iotHubConnectionString, DEFAULT_TOKEN_LIFESPAN_SECONDS);
+    }
+
+    /**
+     * Constructor. Generates a SAS token that grants access to an IoT Hub for
+     * the specified amount of time. (1 year specified in DEFAULT_TOKEN_LIFESPAN_SECONDS)
+     *
+     * @param iotHubConnectionString Connection string object containing the connection parameters.
+     * @param tokenLifespanSeconds The number of seconds that the created SAS token will be valid for.
+     */
+    public IotHubServiceSasToken(IotHubConnectionString iotHubConnectionString, long tokenLifespanSeconds)
+    {
         if (iotHubConnectionString == null)
         {
             throw new IllegalArgumentException();
         }
-        // Codes_SRS_SERVICE_SDK_JAVA_IOTHUBSERVICESASTOKEN_12_002: [The constructor shall create a target uri from the url encoded host name)]
-        // Codes_SRS_SERVICE_SDK_JAVA_IOTHUBSERVICESASTOKEN_12_003: [The constructor shall create a string to sign by concatenating the target uri and the expiry time string (one year)]
-        // Codes_SRS_SERVICE_SDK_JAVA_IOTHUBSERVICESASTOKEN_12_004: [The constructor shall create a key from the shared access key signing with HmacSHA256]
-        // Codes_SRS_SERVICE_SDK_JAVA_IOTHUBSERVICESASTOKEN_12_005: [The constructor shall compute the final signature by url encoding the signed key]
-        // Codes_SRS_SERVICE_SDK_JAVA_IOTHUBSERVICESASTOKEN_12_006: [The constructor shall concatenate the target uri, the signature, the expiry time and the key name using the format: "SharedAccessSignature sr=%s&sig=%s&se=%s&skn=%s"]
+
+        if (tokenLifespanSeconds <= 0)
+        {
+            tokenLifespanSeconds = DEFAULT_TOKEN_LIFESPAN_SECONDS;
+        }
+
+        this.tokenLifespanSeconds = tokenLifespanSeconds;
         this.resourceUri = iotHubConnectionString.getHostName();
         this.keyValue = iotHubConnectionString.getSharedAccessKey();
         this.keyName = iotHubConnectionString.getSharedAccessKeyName();
-        this.expiryTime = buildExpiresOn();
+        this.expiryTimeSeconds = buildExpiresOn();
         this.token =  buildToken();
     }
 
+    /**
+     * @return the number of milliseconds since the UNIX Epoch when this token will expire
+     */
+    public final long getExpiryTimeMillis()
+    {
+        // Multiply by 1000 to convert from seconds to milliseconds
+        return (this.expiryTimeSeconds * 1000);
```

Comment:
would it be better to use something like Duration to signify these time spans, instead of using int/ long? unit conversion like this does increase the scope of bugs

Reply:
A lot of the time logic within this SDK currently uses longs to represent milliseconds like this, so I'm hesitant to change it here. I see some risk in converting from millis since the unix epoch into Duration or Instant types